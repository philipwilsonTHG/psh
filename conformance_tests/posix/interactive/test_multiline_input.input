# Test multiline input handling
# PSH supports multi-line command continuation (v0.26.0 feature)

echo "Testing multiline input handling:"

# Test line continuation with backslash
echo "Testing line continuation:"
echo "This is a long line" \
     "continued on next line"

echo "Math with continuation:" \
     $((1 + 2 + 3))

# Test multiline string literals
echo "Testing multiline strings:"
cat << 'EOF'
This is a multiline
string that spans
multiple lines
EOF

# Test multiline with variable expansion
name="PSH"
cat << EOF
This is $name
on multiple lines
with variable expansion
EOF

# Test multiline commands with proper continuation
echo "Testing multiline command structures:"

# Multiline if statement (typical interactive pattern)
if [ "test" = "test" ]; then
    echo "Multiline if works"
    echo "Second line in if"
fi

# Multiline for loop
for item in one two three; do
    echo "Processing: $item"
    echo "  With additional info"
done

# Multiline while loop
count=1
while [ $count -le 2 ]; do
    echo "Count: $count"
    echo "  In multiline while"
    count=$((count + 1))
done

# Test multiline function definition
multiline_func() {
    echo "This is a multiline"
    echo "function definition"
    local var="local value"
    echo "Local variable: $var"
}

echo "Testing multiline function:"
multiline_func

# Test multiline case statement
value="test"
case "$value" in
    test)
        echo "Case matched"
        echo "Multiple lines in case"
        ;;
    other)
        echo "Other case"
        ;;
    *)
        echo "Default case"
        ;;
esac

# Test multiline pipelines  
echo "Testing multiline pipelines:"
echo "data" | cat | cat

# Test multiline command substitution
echo "Testing multiline command substitution:"
result=$(echo "First line"; echo "Second line")
echo "Result: $result"

# Test multiline arithmetic
echo "Testing multiline arithmetic:"
result=$((1 + 2 + 3 + 4))
echo "Arithmetic result: $result"

# Test multiline variable assignment
echo "Testing multiline variable assignment:"
long_var="This is a long string \
that continues on the next line"
echo "Long variable: $long_var"

# Test multiline array assignment (if supported)
echo "Testing multiline lists:"
list="item1
item2
item3"
echo "Multiline list: $list"

# Test multiline with comments
echo "Testing multiline with comments:"
echo "Command with comment" \
     # This is a comment
     "and continuation"

# Test multiline in different quote contexts
echo "Testing multiline in quotes:"
echo "This is a multiline
string in double quotes"

echo 'This is a multiline
string in single quotes'

# Test complex multiline structure
echo "Testing complex multiline structure:"
for file in $(ls | head -3); do
    if [ -f "$file" ]; then
        echo "File: $file"
        echo "  Type: regular file"
    else
        echo "File: $file"
        echo "  Type: not regular"
    fi
done

# Test multiline error handling
echo "Testing multiline error handling:"
if false; then
    echo "This should not execute"
    echo "Even in multiline context"
else
    echo "Error handling works"
    echo "In multiline else clause"
fi

# Test nested multiline structures
echo "Testing nested multiline structures:"
for outer in 1 2; do
    echo "Outer: $outer"
    for inner in a b; do
        echo "  Inner: $inner"
        if [ "$outer" = "1" ]; then
            echo "    Special case"
        fi
    done
done

echo "Multiline input test completed"