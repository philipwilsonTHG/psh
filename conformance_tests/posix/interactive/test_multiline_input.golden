Testing multiline input handling:
Testing line continuation:
This is a long line continued on next line
Math with continuation: 6
Testing multiline strings:
Testing multiline command structures:
Multiline if works
Second line in if
Processing: one
  With additional info
Processing: two
  With additional info
Processing: three
  With additional info
Count: 1
  In multiline while
Count: 2
  In multiline while
Testing multiline function:
This is a multiline
function definition
Local variable: local value
Case matched
Multiple lines in case
Testing multiline pipelines:
data
Testing multiline command substitution:
Result: First line
Second line
Testing multiline arithmetic:
Arithmetic result: 10
Testing multiline variable assignment:
Long variable: This is a long string that continues on the next line
Testing multiline lists:
<command>:14: Parse error at position 7: Expected delimiter after here document operator
psh: This: command not found
psh: string: command not found
psh: multiple: command not found
psh: EOF: command not found
psh: This: command not found
psh: on: command not found
psh: with: command not found
psh: EOF: command not found
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/Users/pwilson/src/psh/psh/__main__.py", line 169, in <module>
    main()
    ~~~~^^
  File "/Users/pwilson/src/psh/psh/__main__.py", line 163, in main
    exit_code = shell.run_command(script_content, add_to_history=False)
  File "/Users/pwilson/src/psh/psh/shell.py", line 391, in run_command
    return self.script_manager.execute_from_source(input_source, add_to_history)
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/pwilson/src/psh/psh/scripting/base.py", line 48, in execute_from_source
    return self.source_processor.execute_from_source(input_source, add_to_history)
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/pwilson/src/psh/psh/scripting/source_processor.py", line 92, in execute_from_source
    tokens = tokenize(test_command)
  File "/Users/pwilson/src/psh/psh/lexer/__init__.py", line 62, in tokenize
    tokens = lexer.tokenize()
  File "/Users/pwilson/src/psh/psh/lexer/core.py", line 390, in tokenize
    handler()
    ~~~~~~~^^
  File "/Users/pwilson/src/psh/psh/lexer/state_handlers.py", line 95, in handle_double_quote_state
    self._process_quoted_string('"', allow_expansions=True)
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/pwilson/src/psh/psh/lexer/state_handlers.py", line 340, in _process_quoted_string
    self._validate_closing_character(quote_char, error_msg)
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/pwilson/src/psh/psh/lexer/helpers.py", line 345, in _validate_closing_character
    self._error(error_message)
    ~~~~~~~~~~~^^^^^^^^^^^^^^^
  File "/Users/pwilson/src/psh/psh/lexer/core.py", line 330, in _error
    return self.error_handler.handle_error(self, message)
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^
  File "/Users/pwilson/src/psh/psh/lexer/position.py", line 333, in handle_error
    raise error
psh.lexer.position.LexerError:
Lexer Error: Unclosed double quote
  at line 1, column 12

>    1 | list="item1
       |            ^
